### ISSUE TITLE
Joining/Creating job records takes excessively long

### ISSUE THREAD

**Issue by maxfischer2781** [apel/apel#232]
We consistently observe that the database operations to create Job Records take excessively long.
```
Feb 05 19:49:03 htcondor-ce-2-kit condor_ce_apel.sh[1908596]: 2021-02-05 19:49:03,931 - client - INFO - =====================
Feb 05 19:49:03 htcondor-ce-2-kit condor_ce_apel.sh[1908596]: 2021-02-05 19:49:03,931 - client - INFO - Starting joiner.
Feb 05 19:49:03 htcondor-ce-2-kit condor_ce_apel.sh[1908596]: 2021-02-05 19:49:03,931 - apel.db.backends.mysql - INFO - Creating JobRecords
Feb 06 22:01:42 htcondor-ce-2-kit condor_ce_apel.sh[1908596]: 2021-02-06 22:01:42,880 - apel.db.backends.mysql - INFO - Done.
Feb 06 22:01:42 htcondor-ce-2-kit condor_ce_apel.sh[1908596]: 2021-02-06 22:01:42,919 - client - INFO - Joining complete.
Feb 06 22:01:42 htcondor-ce-2-kit condor_ce_apel.sh[1908596]: 2021-02-06 22:01:42,920 - client - INFO - =====================
```
Observed times are up to *and above* 24h. Other DB operations are well below 10s runtime.

See also htcondor/htcondor-ce#419.

**Comment by tofu-rocketry** [comment]
Do you what sort of number of records are being processed? What hardware is this running on?

**Comment by maxfischer2781** [comment]
Sorry for missing the information. Let me know if you need anything more.

20k to 35k jobs per day. Intel Xeon Gold 6226 CPU @ 2.70GHz, 60GiB RAM (~20GiB used). Backend is a local MariaDB with just the APEL database initialised by `schemas/client.sql`.  
We have four machines which all show similar timings.

On our slower (2.20GHz) auxiliary/test machines with <= 6k jobs the join happens in less than 15 minutes.

**Comment by maxfischer2781** [comment]
As a frame of reference, we see a linear scaling on all instances. For the production instances, the scaling is much steeper though.
![apel_scaling_2](https://user-images.githubusercontent.com/5708444/107791630-773b7d80-6d54-11eb-9121-c26c879c83ad.png)
The average gradient is about 800 jobs / h for production, but 23000 jobs / h for cloud and 500k jobs / h for pps/test. I am missing historical data to say whether the gradients degrade over time; obviously the production instances have a lot more total jobs.

Is there any way for the database to accumulate orphan data that could cause such a slowdown? Can we provide you with some database statistics to pinpoint the issue?

**Comment by maxfischer2781** [comment]
Looking at the tables and procedures, it seems that neither the ``BlahdRecords`` table nor the ``JoinJobRecords`` procedure have any measure to skip previously processed BLAH records. Running an ``EXPLAIN`` on the select of ``JoinJobRecords`` shows that ``EventRecords`` use the key ``EventJoinIdx`` but ``BlahdRecords`` *do not* use the key ``BlahdJoinIdx``.

My MySQL is a little rusty, but as far as I can tell that means the *new* ``EventRecords`` are compared against *all* ``BlahdRecords``, without any optimisation for the latter. This would roughly match our observation that machines with more records take longer to process each record.

Is that assumption correct? Is there any means for us to optimise that only new ``BlahdRecords`` are used when matching new ``EventRecords``? For that matter, is there a reason why all ``BlahdRecords`` are used?

**Comment by tofu-rocketry** [comment]
That does sound possible. I'll schedule some time to produce a patch and updated schema. Would you be willing to test it out?

**Comment by maxfischer2781** [comment]
Thanks for looking into this. Yes, I can test a possible fix both on a test and production system.

**Comment by maxfischer2781** [comment]
Do you think it is safe/possible to run the procedure concurrently on the same DB, to process new entries in time? The "real" problem is not necessarily that each run takes long, but that the next run gets delayed and thus has even more data to process. Can we start the "next" ``JoinJobRecords`` while the "current" ``JoinJobRecords`` still runs?

**Comment by tofu-rocketry** [comment]
I've got someone with time to look at this now.

I'd like to try and solve the performance issue first by skipping the already processed fields. That seems the most direct solution.

Possibly related to #20.

**Comment by maxfischer2781** [comment]
Has there been any progress on this? 2 out of our 4 production CE silently entered this state again over the past few weeks.

**Comment by pjcon** [comment]
I ran some more tests which weren't conclusively better in the end. It might be a case of testing it in a live environment rather than a mockup, so in this case I hope to publish some database changes to test on your end this week

**Comment by maxfischer2781** [comment]
@pjcon Yes, I can set aside one of the production machines for testing if you have an optimised query.

**Comment by maxfischer2781** [comment]
If anyone else needs a workaround:

As a side-effect of #269, we've increased the ``[mysqld] innodb_buffer_pool_size`` to work around another issue. This seems to also speed up JobRecord joining. Both the original 128MB and new 1GB setting seem a bit excessive for joining about 1000 records, but it should keep accounting operational for a while.

**Comment by pjcon** [comment]
@maxfischer2781 

https://github.com/apel/apel/pull/270

I expect your SQL is better than mine, I'd be grateful for any corrections and suggestions!

**Comment by maxfischer2781** [comment]
@pjcon Much appreciated. I've had a few delays, but should be able to test it over the next days.
